### PR TITLE
Separate screen-, workflow- and usb-related functions

### PR DESCRIPTION
**Pull request by conte91**
Moved the usb processing and is_done() logic outside of `_screen_process()`.

Only the main loop should be responsible for calling the USB handlers - does this have anything to do with the GUI after all? Same for the workflow-related logic (blocking, timeouts, etc).

Note that this is just code wrangling so the new implementation should be 100% equivalent to the current one - if it isn't, that's a bug.

Depends on #147.

### CODE REVIEW

File: src/firmware_main_loop.c [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#include "firmware_main_loop.h"
```
Comment (benma) [inline src/firmware_main_loop.c]: license header missing

File: src/ui/screen_process.c [hunk 2]
```diff
@@ -34,6 +31,23 @@ void ui_screen_render_component(component_t* component)
...
 }
 
+/**
+ * Detects if the screen component being displayed has changed
+ * since the last time this function was called.
+ * This stores the last observed component into a global.
+ *
+ * @param[in] current_comp Current on-screen component.
+ */
+static bool _screen_has_changed(const component_t* current_comp)
```
Comment (benma) [inline src/ui/screen_process.c]: please no short names, as explicit long names are easier to understand, is less ambiguous etc.. `current_component`

File: src/ui/screen_process.c [hunk 3]
```diff
@@ -46,52 +60,44 @@ static component_t* _get_waiting_screen(void)
...
+        return _get_waiting_screen();
+    }
+    return result;
 }
 
-void ui_screen_process_with_timeout(
-    bool (*is_done)(void),
-    void (*on_timeout)(void),
-    uint32_t timeout)
+void ui_spin(void)
```
Comment (benma) [inline src/ui/screen_process.c]: this function should be called `screen_process_spin` (usually same as filename).

can rename the whole file and the functions in here in a separate PR, but would keep `screen_process_*` now to keep consistent.

furthermore, `spin` to me means it is looping, but it is not looping. Please find a better name (`screen_process()` ?)

File: src/ui/screen_process.c [hunk 4]
```diff
@@ -46,26 +62,39 @@ static component_t* _get_waiting_screen(void)
     return waiting_screen;
 }
 
-static void _screen_process(bool (*is_done)(void), void (*on_timeout)(void), const uint32_t timeout)
+/**
+ * \param is_done WTF
```
Comment (benma) [inline src/ui/screen_process.c]: :thinking: 

not sure this comment helps  :laughing:

**Comment by benma** [comment]
This looks like a very good cleanup. Code looks good to me!

I will test it some more tomorrow.

File: src/ui/screen_process.c (hunk 3 above)
Comment (conte91) [reply src/ui/screen_process.c]: Fixed. `screen_process` wins.

File: src/ui/screen_process.c (hunk 2 above)
Comment (conte91) [reply src/ui/screen_process.c]: Fixed

File: src/ui/screen_process.c (hunk 4 above)
Comment (conte91) [reply src/ui/screen_process.c]: Comment? Which comment? :D

**Pull request merged**